File: 백준/Silver/20920. 영단어 암기는 괴로워/영단어 암기는 괴로워.py
import sys
from collections import defaultdict
from operator import itemgetter

N, M = map(int, sys.stdin.readline().split())
words = defaultdict(int)
for _ in range(N):
    word = sys.stdin.readline().strip()
    if len(word) >= M:
        words[word] += 1
        
word_list = [(k,v) for k, v in words.items()]
word_list.sort(key=itemgetter(0))
word_list.sort(key=lambda x: len(x[0]), reverse=True)
word_list.sort(key=itemgetter(1),reverse=True)

for word, _ in word_list:
    print(word)

File: 백준/Silver/20920. 영단어 암기는 괴로워/영단어 암기는 괴로워.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"sort"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	writer := bufio.NewWriter((os.Stdout))
	defer writer.Flush()

	// 1 <= N <= 100,000, 1 <= M <= 10
	var n, m int
	fmt.Fscanln(reader, &n, &m)

	// 일단 map 에 저장
	dict := make(map[string]int)

	for i := 0; i < n; i++ {
		var word string
		fmt.Fscanln(reader, &word)

		if len(word) >= m {
			dict[word]++
		}
	}

	type Word struct {
		word string
		count int
	}

	// slice 에 복사
	word_list := []Word{}
	for key, value := range dict {
		word_list = append(word_list, Word{key, value,})
	}

	// 적당한 조건으로 sort
	sort.Slice(word_list, func(i, j int) bool {
		if word_list[i].count == word_list[j].count {
			if len(word_list[i].word) == len(word_list[j].word) {
				return word_list[i].word < word_list[j].word
			}
			return len(word_list[i].word) > len(word_list[j].word)
		}
		return word_list[i].count > word_list[j].count
	})

	// 결과 출력
	for _, word := range word_list {
		fmt.Fprintf(writer, "%s\n", word.word)
	}
}
